File: sem4/jpp/list3/cpp/main.cpp
#include "DHSetup.hpp"
#include "GF1234567891Field.hpp"
#include "User.hpp"

#include <iostream>

int main() {
    DHSetup<GF1234567891Field> dhSetup;
    std::cout << "Generator: " << dhSetup.getGenerator() << '\n';

    User<GF1234567891Field> alice(dhSetup);
    User<GF1234567891Field> bob(dhSetup);
    GF1234567891Field alicePublicKey = alice.getPublicKey();
    GF1234567891Field bobPublicKey = bob.getPublicKey();
    std::cout << "Alice's public key: " << alicePublicKey << '\n';
    std::cout << "Bob's public key: " << bobPublicKey << '\n';

    alice.setKey(bobPublicKey);
    bob.setKey(alicePublicKey);

    GF1234567891Field message = 123;
    GF1234567891Field encryptedMessage = alice.encrypt(message);
    GF1234567891Field decryptedMessage = bob.decrypt(encryptedMessage);

    std::cout << "Original Message: " << message << '\n';
    std::cout << "Encrypted Message: " << encryptedMessage << '\n';
    std::cout << "Decrypted Message: " << decryptedMessage << '\n';

    return 0;
}

File: sem4/jpp/list3/python/main.py
from gf1234567891field import GF1234567891Field
from dh_setup import DHSetup
from user import User


def main() -> None:
    dh_setup = DHSetup(GF1234567891Field)
    print(f"Generator: {dh_setup.generator}")

    alice = User(dh_setup)
    bob = User(dh_setup)
    alice_public_key = alice.public_key
    bob_public_key = bob.public_key
    print(f"Alice's public key: {alice_public_key}")
    print(f"Bob's public key: {bob_public_key}")

    alice.set_key(bob_public_key)
    bob.set_key(alice_public_key)

    message = GF1234567891Field(123)
    encryptedMessage = alice.encrypt(message)
    decryptedMessage = bob.decrypt(encryptedMessage)

    print(f"Original Message: {message}")
    print(f"Encrypted Message: {encryptedMessage}")
    print(f"Decrypted Message: {decryptedMessage}")


if __name__ == "__main__":
    main()
